[CALCITE-4012] Implement trait propagation for EnumerableHashJoin and…

amaliujia opened the pull request:
… EnumerableNestedLoopJoin

amaliujia commented on core/src/test/resources/org/apache/calcite/test/TopDownOptTest.xml:
```diff
@@ -648,4 +648,80 @@ EnumerableMergeJoin(condition=[AND(=($1, $5), =($0, $4))], joinType=[inner])
+EnumerableHashJoin(condition=[AND(=($1, $9), =($2, $10))], joinType=[left])
+  EnumerableSort(sort0=[$2], sort1=[$1], dir0=[DESC-nulls-last], dir1=[ASC-nulls-first])
+    EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
+    EnumerableTableScan(table=[[CATALOG, SALES, BONUS]])
```
@hsyuan 

I want to see this plan (sort on left join input for this left outer join), however its cost is too high(`{67.94680261461362 rows, 1346.0848941260901 cpu, 0.0 io}`)

So Calcite always chooses 
EnumerableSort(sort0=[$2], sort1=[$1], dir0=[DESC-nulls-last], dir1=[ASC-nulls-first])
  EnumerableHashJoin(condition=[AND(=($1, $9), =($2, $10))], joinType=[left])
    EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
    EnumerableTableScan(table=[[CATALOG, SALES, BONUS]])

whose cost is `{54.94680261461362 rows, 68.0 cpu, 0.0 io}`.

It could be make sense if EnumerableHashJoin reduces number of tuples thus Sort at the end could have less cost to sort joined result.


Do you have a suggestion how to deterministically test trait pass down for EnumerableHashJoin?

hsyuan replied:
Don't test it directly. Try to add a aggregate in the left child.

amaliujia replied:
I am testing trait pass through so adding aggregation to child does not solve the problem.

I tried to dig into the cost computation processing. It turns out that for this plan:

```
EnumerableSort(sort0=[$2], sort1=[$1], dir0=[DESC-nulls-last], dir1=[ASC-nulls-first])
  EnumerableHashJoin(condition=[AND(=($1, $9), =($2, $10))], joinType=[left])
    EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
    EnumerableTableScan(table=[[CATALOG, SALES, BONUS]])
```

The top Sort's row count is just 2.1, which was is very surprising cause my debug point at the `computeSelfCost` in `EnumerableHashJoin` hasn't caught 2.1 row count, but ~50 row counts.

As a comparison, the plan I am expecting:
```
EnumerableHashJoin(condition=[AND(=($1, $9), =($2, $10))], joinType=[left])
  EnumerableSort(sort0=[$2], sort1=[$1], dir0=[DESC-nulls-last], dir1=[ASC-nulls-first])
    EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
    EnumerableTableScan(table=[[CATALOG, SALES, BONUS]])
```
The Sort gets 14 row counts, which is from `EnumerableTableScan(table=[[CATALOG, SALES, EMP]])`, which is correct, and that's why this Sort's sort is update to `1346.0848941260901`, which is understandable.


So either I didn't fully understand how Calcite propagate row counts, or something wrong for top-down opt code (e.g. all debug point's stops didn't give me a 2.1 row count at EnumerableHashJoin while the EnumerableSort on top of HashJoin gets that 2.1 row count).

Will continue investigating the root cause.

amaliujia replied:
Ok after trace down the optimizer, I realize that Calcite does an estimated row count computation at [1]. in the estimated rowcount, cause of formula of how JOIN estimated row count is computed, the input for EnumerableSort becomes as low as 2.1.

amaliujia replied:
The estimated cost (from `computeSelfCost`) is certainly different from `bestCost` in `RelSubset` though.

amaliujia replied:
Figured out a way to generate expected plan. See my new comment.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,53 @@ public static EnumerableHashJoin create(
+  }
+
+//  @Override public DeriveMode getDeriveMode() {
+//    return DeriveMode.BOTH;
```
should prohibit for full outer join.

amaliujia replied:
Sounds good for full outer join (will use `DeriveMode.PROHIBITED` to dis-allow it)

Still understanding why RIGHT OUTER JOIN should be forbidon (why EnumerableHashJoin always builds hash table on right input).

hsyuan replied:
I am a little bit skeptical about it. It shouldn't work that way. If there isn't a match, it should generate nulls on left, but building hash table on right side will make it impossible. Can you write a query and execute it?

amaliujia replied:
Will do.

amaliujia replied:
I confirmed that indeed EnumerableHashJoin always builds hash table on right side. 

For right/full outer join, to make result right, the implementation keeps a set of keys from right side. During probing it also removes keys from that set. At the end it will know how many keys from right side does not have a match with left side, thus it can produces null with right tuples. 


So I will prohibit traits propogation for right/full outer join.

hsyuan replied:
Cool, thanks for checking.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,53 @@ public static EnumerableHashJoin create(
+    // It is because for a hash-based implementation, only non-hashed side can
+    // preserve ordering, thus only for left/right outer join, we are sure which
+    // side is non-hashed side (the outer child).
+    if (joinType == JoinRelType.LEFT) {
```
INNER join, SEMI/ANTI SEMI, LEFT all can pass request to left child, and derive from left child.

hsyuan replied:
Only right outer join uses right child as the probe side, I guess. Need to check the implementation of HashJoin.

hsyuan replied:
On the other hand, it shows exactly that trait propagation should only be done on physical operators, because it is physical implementation dependent.

amaliujia replied:
Agreed. I will check EnumerableHashJoin. Basically collation passes down will depend on which side to build the hash table.

rubenada replied:
Please note that currently EnumerableHashJoin implementation builds the hash table always on the right input.

amaliujia replied:
Thanks @rubenada for the reminder. It's surprising though. Let me check the implementation of EnumerableHashJoin to see how it make right to build hash table on right input for **RIGHT OUTER JOIN**.

hsyuan replied:
Then for FULL/RIGHT OUTER JOIN, we should prohibit trait propagations.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,53 @@ public static EnumerableHashJoin create(
+      // Only consider exact key match for now
+      if (requiredKeySet.equals(leftKeySet)) {
+        return Pair.of(
+            required,
```
should not use required directly, because required may have different convention. Itself and children's traits should all have EnumerableConvention. See how EnumerableFilter does.

amaliujia commented on core/src/test/java/org/apache/calcite/test/TopDownOptTest.java:
```diff
@@ -288,6 +290,44 @@
+        + "(select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp) r left outer join\n"
+        + "(select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bon ...
+        + "on r.ename=s.ename\n"
+        + "order by r.ename nulls first";
```
@hsyuan 
Finally I find a way to "fool" planner to choose the plan I want, basically the plan that pushes sort down to left child of hash join.

The reason that simple join's cheapest plan has a Sort on top of HashJoin, is because the Join's estimated is computed as simple as `left_row_count * right_row_count * 0.15`, 0.15 is selectively. When inputs's row count is small, the selectively plays a bigger role, thus HashJoin gives a small cost, thus Sort is put after Join because Join gives a very small row count estimation.

So I intentionally use UNION to increase input size, which make the  ``left_row_count * right_row_count` plays a bigger role for Join, and then Join produces larger row count estimation. Then finally planner believes push down Sort is cheaper, thus gives the plan I want to see.

amaliujia replied:
I know the test query becomes ugly and not easy to check, but I don't know if there is a better idea.

amaliujia replied:
It's best actually if current testing suites can allow us mock data, due to cost-based model, its's better to use data rather than wired query to play with it.

I could also check if I can modify the testing framework.

hsyuan replied:
How about this:
```sql
select * from
(select distinct a, b from foo) foo
join bar on foo.c = bar.c
order by b desc, a desc; 
```

amaliujia replied:
Add an aggregation is not helping. The hash join still uses `left_row_count * right_row_count * selectivity` to estimate cost while the aggreagtion cannot give a big enough `left_row_count`.

amaliujia replied:
Problem is solved. I noticed that there are several large tables in `customers` schema. Those large tables can generate plans I want.And if there are future needs, we can always mock new tables for optimization testing.

I pushed a new commit to illustrate which table I am using.

With good input data, I will continue implementation work.

amaliujia commented:
@hsyuan 

I have push commits for collation pushdown for both EnumerableHashJoin and EnumerableNestedLoopJoin.

Regarding traits derivation, I gave some thoughts on it and now I think it is not useful. It is because for both EnumerableHashJoin and EnumerableNestedLoopJoin, they can perserve ordering from left join input. Based on examples of MergeJoin, seems like the goal of trait derivation was to enforce a sort on right input, which does not work for hashjoin and nested loop join in this case.

hsyuan commented:
> Based on examples of MergeJoin, seems like the goal of trait derivation was to enforce a sort on right input, which does not work for hashjoin and nested loop join in this case.

This is incorrect. `derive()` is used to derived new traits from child trait, just like project, filter. MergeJoin is different, because it requires both sides sorted on the same keys.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,45 @@ public static EnumerableHashJoin create(
+              passthroughTraitSet.replace(RelCollations.EMPTY)));
+    }
+
+    // TODO: consider super keyset and sub keyset
```
HashJoin/NestloopJoin doesn't need this.

amaliujia replied:
How about this query in which it uses HashJoin:

 "select * from\n"
        + "(select contactno, email from customer.contact_peek) r left outer join\n"
        + "(select acctno, type from customer.account) s\n"
        + "on r.contactno=s.acctno and r.email=s.type\n"
        + "order by r.contactno desc";

I think we can still pass collation `r.contactno desc`?

hsyuan replied:
Correct. But might not be selected due to cost reason.

hsyuan replied:
so this comment should be removed.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,45 @@ public static EnumerableHashJoin create(
+      return DeriveMode.PROHIBITED;
+    }
+
+    return DeriveMode.BOTH;
```
Unless you want to support distribution, otherwise, `LEFT_FIRST` should be the only choice.

amaliujia replied:
Sure. let me make a change.  I haven't fully understood the differnet of `LEFT_FIRST`, `RIGHT_FIRST` and `BOTH`.

hsyuan replied:
I need to consider refining the java doc to make it easier to understand.

amaliujia commented on core/src/test/java/org/apache/calcite/test/TopDownOptTest.java:
```diff
@@ -287,6 +289,119 @@
   }
+
+  // Not push down sort for hash join in inner join case.
+  @Test void testHashJoinInnerJoinNotPushDownSort() {
```
In fact, this test case is not valid (sort not pushdown due to cost-baed optimization).

I will remove this one.

amaliujia commented:
Ok then. I can go ahead to support trait derivation.

amaliujia commented:
@hsyuan any suggestions on how hash join traits derivation should be tested?

I am thinking doing this
```
select * from
(select distinct a, b from foo) foo
join bar on foo.c = bar.c 
```

Then the produced plan might add a Sort on top of Select.

Do you think it is right?

hsyuan commented:
How about this:
```sql
select * from (select a,b,c from foo oder by a desc, b desc limit 10) foo
join bar on foo.a = bar.a
order by foo.a desc, foo.b desc;
```

amaliujia commented:
@hsyuan 
Thanks for your suggestion. Added traits derivation and properly tests.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,81 @@ public static EnumerableHashJoin create(
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> deriveTraits(
+      final RelTraitSet childTraits, final int childId) {
+    // should only derive traits (limited to collation for now) from left join input.
+    if (childId > 0) {
```
should be `assert childId == 0`, it is illegal to be greater than 0

amaliujia replied:
Ah yes. I think it is because this code block

```
        if (mode == DeriveMode.LEFT_FIRST
            || mode == DeriveMode.RIGHT_FIRST) {
          break;
        }
```

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,81 @@ public static EnumerableHashJoin create(
+    // and only non-hashed side can preserve ordering.
+
+    // Only consider exact key match for now
+    if (requiredKeySet.equals(leftKeySet)) {
```
hash join / nested loopjoin are different with MergeJoin. The required collation keys can be any keys other than join keys. 
```sql
select * from foo join bar using (a) order by foo.b;
```
We can still pass down the collation request to HashJoin's left child, because hashjoin doesn't require any collation from its left child.
But for mergejoin, we can't pass down. because the join's parent requires collation on `b`, but the mergejoin requires its children to be sorted on `a`.

amaliujia replied:
This is a greta point.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,81 @@ public static EnumerableHashJoin create(
+    final int keyCount = joinInfo.leftKeys.size();
+    if (colCount < keyCount || keyCount == 0) {
+      return null;
+    }
```
this is not needed. we can derive any collation from its left child. no matter what are the join keys.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,81 @@ public static EnumerableHashJoin create(
+
+    if (colCount > keyCount) {
+      collation = RelCollations.of(collation.getFieldCollations().subList(0, keyCount));
+    }
```
same here.

amaliujia commented:
It is a great point that HashJoin does not care about collations, thus it can pass through or derive collations. Updated this PR and also added a few test cases.

amaliujia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableNestedLoopJoin.java:
```diff
@@ -118,6 +129,104 @@ public static EnumerableNestedLoopJoin create(
+    return DeriveMode.LEFT_FIRST;
+  }
+
+  private Set<Integer> extractTrivialFieldOfLeftInputFromJoinCondition() {
```
In fact, I should also change NestedLoopJoin because it also does not require to match with join keys.

hsyuan replied:
Correct.

amaliujia replied:
Done. Updated NestedLoopJoin with several extra tests.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableNestedLoopJoin.java:
```diff
@@ -118,6 +124,64 @@ public static EnumerableNestedLoopJoin create(
+
+    return Pair.of(
+        getTraitSet().replace(collation),
+        ImmutableList.of(childTraits, right.getTraitSet()));
```
We also need to replace childTraits with enumerable convention. also hashjoin.

amaliujia replied:
Done.

amaliujia commented:
Also rebased to master and squashed commits to one commit with better description.

hsyuan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java:
```diff
@@ -100,6 +106,57 @@ public static EnumerableHashJoin create(
+    RelCollation collation = childTraits.getCollation();
+    if (collation == null
+        || collation == RelCollations.EMPTY
+        || joinType == JoinRelType.FULL
+        || joinType == JoinRelType.RIGHT) {
```
and here

amaliujia replied:
Done

hsyuan commented on core/src/test/resources/org/apache/calcite/test/TopDownOptTest.xml:
```diff
@@ -648,4 +648,472 @@ EnumerableMergeJoin(condition=[AND(=($1, $5), =($0, $4))], joinType=[inner])
+    EnumerableProject(CONTACTNO=[$0], FNAME=[$1], LNAME=[$2], EMAIL=[$3], X=[$4.X], Y=[$4.Y], unit=[$4.unit], M=[$5.M], A=[$5.SUB.A], B=[$5.SUB.B])
+      EnumerableSort(sort0=[$0], sort1=[$3], dir0=[DESC], dir1=[DESC])
+        EnumerableTableScan(table=[[CATALOG, CUSTOMER, CONTACT_PEEK]])
+    EnumerableTableScan(table=[[CATALOG, CUSTOMER, ACCOUNT]])
```
Left outer join's cardinality is always greater equal with its left input's cardinality. So this plan is always better than sorting on join's output. 👍

amaliujia commented on core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableNestedLoopJoin.java:
```diff
@@ -118,6 +124,65 @@ public static EnumerableNestedLoopJoin create(
+    if (collation == null
+        || collation == RelCollations.EMPTY
+        || joinType == JoinRelType.FULL
+        || joinType == JoinRelType.RIGHT) {
```
Ah I see. The deriveMode method is used for traits derivation.

amaliujia commented:
Comments addressed! Thank you!

hsyuan commented:
@amaliujia Some tests in MaterializedViewRelOptRulesTest failed to match MV, can you take a look why?

amaliujia commented:
You mean to test top-down opt globally? Will take a look.

amaliujia commented:
@hsyuan 

Have checked failed MaterializedViewRelOptRulesTest after enabled top-down opt globally.

There were three tests:
testJoinAggregateMaterializationNoAggregateFuncs9()
testJoinAggregateMaterializationAggregateFuncs11()
testJoinMaterialization10()


These failed tests failed as expected, because of this PR.

Basically in the original tests, they usually generate:
`EnumerableCalc(expr#0..3=[{inputs}], expr#4=[CAST($t1):VARCHAR], expr#5=[10], expr#6=[>($t0, $t5)], expr#7=[30], expr#8=[>=($t7, $t0)], expr#9=[AND($t6, $t8)], deptno=[$t0], name0=[$t4], $condition=[$t9])`, 

and then test tried to find
`expr#5=[1], expr#6=[>($t0, $t5)], expr#7=[30], expr#8=[>=($t7, $t0)]`.

Because sort is pushed down, the original Calc is split to two
```
          EnumerableCalc(expr#0..3=[{inputs}], expr#4=[CAST($t1):VARCHAR], deptno=[$t0], name0=[$t4])
            EnumerableSort(sort0=[$0], dir0=[ASC])
              EnumerableCalc(expr#0..3=[{inputs}], expr#4=[10], expr#5=[>($t0, $t4)], expr#6=[30], expr#7=[>=($t6, $t0)], expr#8=[AND($t5, $t7)], proj#0..3=[{exprs}], $condition=[$t8])
```


As you can see, CAST stays in the first Calc, thus the second Calc does not have the CAST, and some reference offset are shifted by 1.

So the test will need to find `expr#4=[10], expr#5=[>($t0, $t4)], expr#6=[30], expr#7=[>=($t6, $t0)]`.

amaliujia commented:
The failed test in MaterializedViewRelOptRulesTest after enable top-down opt reminds me that, it is still not a right time to enable top-down opt globally, because top-down opt development still on-going. 

Some tests are easy to be affected by top-down opt.

hsyuan commented:
It is disappointing that the Calcite's MV is trying to match physical plans. Further more, Choosing MV or not based on cost is totally meaningless.
